File: 3525-maximum-energy-boost-from-two-drinks/maximum-energy-boost-from-two-drinks.cpp
class Solution {
public:
    int n;
    long long dp[100001][3];
    long long solve(int index,int drinkA,vector<int>& energyDrinkA, vector<int>& energyDrinkB)
    {
        if(index>=n)
        return 0;

        if(dp[index][drinkA]!=-1)
        return dp[index][drinkA];

        long long maxEnergy=0;
        if(drinkA)
        {
            // drink A and not switch
            long long A1=energyDrinkA[index]+solve(index+1,1,energyDrinkA,energyDrinkB);
            // drink A and switch
            long long A2=energyDrinkA[index]+solve(index+2,0,energyDrinkA,energyDrinkB);
            maxEnergy=max(A1,A2);
        }
        else
        {
            // drink B and not switch
            long long B1=energyDrinkB[index]+solve(index+1,0,energyDrinkA,energyDrinkB);
            // drink B and switch
            long long B2=energyDrinkB[index]+solve(index+2,1,energyDrinkA,energyDrinkB);
            maxEnergy=max(B1,B2);
        } 
        return dp[index][drinkA]=maxEnergy;
    }

    long long maxEnergyBoost(vector<int>& energyDrinkA, vector<int>& energyDrinkB) 
    {
        n=energyDrinkA.size();
        memset(dp,-1,sizeof(dp));
        long long startWithA=solve(0,1,energyDrinkA,energyDrinkB);
        long long startWithB=solve(0,0,energyDrinkA,energyDrinkB);
        return max(startWithA,startWithB);
    }
};

File: 3525-maximum-energy-boost-from-two-drinks/maximum-energy-boost-from-two-drinks.java
class Solution {
    private Long[][] dp;
    private long solve(int index, int currDrink, int[] energyDrinkA, int[] energyDrinkB)
    {
        if(index >= energyDrinkA.length)
        return 0;

        if(dp[index][currDrink] != null)
        return dp[index][currDrink];

        long drinkBoost = currDrink == 0 ? energyDrinkA[index] : energyDrinkB[index];
        
        // two choices
        long notSwitch = solve(index + 1, currDrink, energyDrinkA, energyDrinkB);
        long switchDrink = solve(index + 2, 1 - currDrink, energyDrinkA, energyDrinkB);
        
        return dp[index][currDrink] = drinkBoost + Math.max(notSwitch, switchDrink);
    }

    public long maxEnergyBoost(int[] energyDrinkA, int[] energyDrinkB) 
    {
        dp = new Long[energyDrinkA.length][2];
        return Math.max(solve(0, 0, energyDrinkA, energyDrinkB), solve(0, 1, energyDrinkA, energyDrinkB));
    }
}